frontend/src/context/dashboardStateSlice.ts

```diff
@@ -34,18 +169,219 @@ export const dashboardStateSlice = createSlice({
+        keepLayer(l, layer),
+      );
+
+      return {
+        ...state,
+        maps: state.maps.map((mapInstance, i) =>
+          i === index
+            ? { ...mapInstance, layers: filteredLayers }
+            : mapInstance,
+        ),
+      };
+    },
+    updateMapDateRange: (
+      state,
+      action: PayloadAction<{ index: number; dateRange: DateRange }>,
+    ) => {
+      const { index, dateRange } = action.payload;
+
+      return {
+        ...state,
+        maps: state.maps.map((mapInstance, i) =>
+          i === index ? { ...mapInstance, dateRange } : mapInstance,
+        ),
+      };
+    },
+    setMap: (
+      state,
+      action: PayloadAction<{ index: number; maplibreMap: MapGetter }>,
+    ) => {
+      const { index, maplibreMap } = action.payload;
+      return {
+        ...state,
+        maps: state.maps.map((mapInstance, i) =>
+          i === index ? { ...mapInstance, maplibreMap } : mapInstance,
+        ),
+      };
+    },
+    removeLayerData: (
+      state,
+      action: PayloadAction<{ index: number; layer: LayerType }>,
+    ) => {
+      const { index, layer } = action.payload;
+      return {
+        ...state,
+        maps: state.maps.map((mapInstance, i) =>
+          i === index
+            ? {
+                ...mapInstance,
+                layersData: mapInstance.layersData.filter(
+                  ({ layer: dataLayer }) => dataLayer.id !== layer.id,
+                ),
+              }
+            : mapInstance,
+        ),
+      };
+    },
+    setBoundaryRelationData: (
+      state,
+      action: PayloadAction<{ index: number; data: BoundaryRelationsDict }>,
+    ) => {
+      const { index, data } = action.payload;
+      return {
+        ...state,
+        maps: state.maps.map((mapInstance, i) =>
+          i === index
+            ? { ...mapInstance, boundaryRelationData: data }
+            : mapInstance,
+        ),
+      };
+    },
+    dismissError: (
+      state,
+      action: PayloadAction<{ index: number; error: string }>,
+    ) => {
+      const { index, error } = action.payload;
+      return {
+        ...state,
+        maps: state.maps.map((mapInstance, i) =>
+          i === index
+            ? {
+                ...mapInstance,
+                errors: mapInstance.errors.filter(msg => msg !== error),
+              }
+            : mapInstance,
+        ),
+      };
+    },
+    setDashboardOpacity: (
+      state,
+      action: PayloadAction<{ index: number } & SetDashboardOpacityParams>,
+    ) => {
+      const { index, map, layerId, layerType, value, callback } =
+        action.payload;
+
+      if (!map || !layerId || value === undefined || !layerType) {
+        return state;
+      }
+
+      const [mapLayerId, opacityType] = getMapLayerOpacityConfig(
+        layerId,
+        layerType,
+      );
+
+      // update map
+      if (map.getLayer(mapLayerId) !== undefined && value !== undefined) {
```

Comment:
Looks like a copy/paste from another opacity manager. Worth deduping? Maybe the hacky part?